### PR TITLE
unite electra pretrain, fine-tune and deploy

### PR DESCRIPTION
**Pull request by jeff41404**
1. electra模型支持全流程：预训练、fine-tuning和推理部署，并无缝衔接；
2. electra模型支持混合精度(float16)训练、支持一键切换float64训练；
3. electra模型支持断点恢复训练；
4. electra模型组网动静统一、支持动态图转静态图，支持静态图推理部署

### CODE REVIEW

File: paddlenlp/transformers/electra/tokenizer.py [hunk 1]
```diff
@@ -121,7 +121,7 @@ def _tokenize(self, text):
                 split_tokens.append(sub_token)
         return split_tokens
 
-    def __call__(self, text):
+    def tokenize(self, text):
```
Comment (smallv0221) [inline paddlenlp/transformers/electra/tokenizer.py:125]: 还有一个function需要接入，build_offset_mapping_with_special_tokens()，参考BertTokenizer。

File: paddlenlp/transformers/electra/tokenizer.py [hunk 2]
```diff
@@ -121,7 +121,7 @@ def _tokenize(self, text):
...
         """
```
Comment (smallv0221) [inline paddlenlp/transformers/electra/tokenizer.py:126]: 删除encode()方法，这个方法也继承PretrainedTokenizer就行。

File: paddlenlp/transformers/electra/tokenizer.py [hunk 3]
```diff
@@ -121,7 +121,7 @@ def _tokenize(self, text):
...
         End-to-end tokenization for Electra models.
```
Comment (smallv0221) [inline paddlenlp/transformers/electra/tokenizer.py:127]: 注释里面有几处写的还是BERT，是不是需要改一下

File: paddlenlp/transformers/electra/tokenizer.py (hunk 1 above)
Comment (jeff41404) [reply paddlenlp/transformers/electra/tokenizer.py:125]: 已加上

File: paddlenlp/transformers/electra/tokenizer.py (hunk 2 above)
Comment (jeff41404) [reply paddlenlp/transformers/electra/tokenizer.py:126]: 已删除

File: paddlenlp/transformers/electra/tokenizer.py (hunk 3 above)
Comment (jeff41404) [reply paddlenlp/transformers/electra/tokenizer.py:127]: 之前好像有同学改过，已修改

File: examples/language_model/electra/README.md [hunk 4]
```diff
@@ -1,53 +1,143 @@
...
+```shell
+python -u ./get_ft_model.py \
+    --model_dir ./pretrain_model/model_40000.pdparams/
+```
+其中参数释义如下：
+- `model_dir` 表示预训练模型所在目录，这里例子取预训练40000步的checkport来生成Fine-tuning所需模型，生成的模型也会在这个目录下。
+此命令可多次执行，但只有第1次会生成Fine-tuning所需模型
+
+### 4.2 运行Fine-tuning
+使用../glue/run_glue.py运行，详细可参考../glue/README.md，有两种方式：
```
Comment (ZeyuChen) [inline examples/language_model/electra/README.md]: 这里的路径指引好像不太准确

File: paddlenlp/transformers/electra/tokenizer.py [hunk 5]
```diff
@@ -121,7 +121,7 @@ def _tokenize(self, text):
                 split_tokens.append(sub_token)
         return split_tokens
```
Comment (smallv0221) [inline paddlenlp/transformers/electra/tokenizer.py:124]: 辛苦在最上面在加一下HuggingFace的license，也参考BertTokenizer

File: examples/language_model/electra/README.md [hunk 6]
```diff
@@ -58,19 +148,44 @@ cd ../glue/ && python -u ./run_glue.py \
 - `output_dir` 表示模型保存路径。
 - `n_gpu` 表示使用的 GPU 卡数。若希望使用多卡训练，将其设置为指定数目即可；若为0，则使用CPU。
 
-训练过程将按照 `logging_steps` 和 `save_steps` 的设置打印如下日志：
+#### 4.2.2 使用本地预训练模型运行 Fine-tuning
+按照上面第3章在本地运行 ELECTRA 模型的预训练后，执行4.1的命令得到Fine-tuning所需模型，然后运行 Fine-tuning。
+
+以 GLUE/SST-2 任务为例，启动 Fine-tuning 的方式如下（`paddlenlp` 要已经安装或能在 `PYTHONPATH` 中找到）：
```
Comment (ZeyuChen) [inline examples/language_model/electra/README.md]: `paddlenlp` 要已经安装或能在 `PYTHONPATH` 中找到
应该不用强调这点

File: examples/language_model/electra/README.md [hunk 7]
```diff
@@ -85,28 +200,78 @@ eval loss: 0.549763, acc: 0.9151376146788991, eval done total : 1.82069873809814
...
-例如[Gutenberg Dataset](https://web.eecs.umich.edu/~lahiri/gutenberg_dataset.html)
-下面例子假设数据在./BookCorpus/，数据文件为纯文本train.data
 
+## 5. 推理部署
+### 5.1 导出推理模型
 ```shell
-export CUDA_VISIBLE_DEVICES=0,1
-export DATA_DIR=./BookCorpus/
+python -u ./export_model.py \
+    --input_model_dir ./pretrain_model/model_40000.pdparams/ \
```
Comment (ZeyuChen) [inline examples/language_model/electra/README.md]: 部署导出的模型，是否应该以--output_dir下finetune后的模型产出为示例比较合理。而不是以与训练阶段的参数。因为部署都是针对下游finetune的模型进行部署

File: examples/language_model/electra/README.md [hunk 8]
```diff
@@ -85,28 +200,78 @@ eval loss: 0.549763, acc: 0.9151376146788991, eval done total : 1.82069873809814
...
+    --params_file ../../electra-small.pdiparams \
+    --predict_file "../../test.txt" "../../test.txt.1" \
+    --batch_size 1 \
+    --max_seq_length 128 \
+    --model_name electra-small
+```
+其中绝大部分和从命令行读取输入数据一样，这里描述不一样的参数：
+- `predict_file` 表示用于推理的文件数据，可以配置1个或多个文件，每个文件和2.2预训练数据格式一样，为utf-8编码的文本数据，每行1句话。如果此项配置，则predict_sentences不用配置。
+
+对于每1句话模型推理分别给出1个推理结果，即Discriminator的Replaced Token Detection(RTD)任务结果，是由0和1组成的list，list长度等于句子长度加2（因推理时会在每句话首尾添加特殊符号[CLS]和[SEP]，和训练时的数据处理一致），0表示句子中对应位置上的单词没有被替换过（是原文），1表示句子中对应位置上的单词是被替换过的（不是原文&不符合语言逻辑）。
```
Comment (ZeyuChen) [inline examples/language_model/electra/README.md]: 感觉以这个场景作为ELECTRA的部署示例不太准确，因为用户需要的是针对下游的finetune任务
如文本分类，序列标注等，以ELECTRA判别的输出作为推理部署不是一个主流的用户用法。

File: examples/language_model/electra/deploy/python/predict.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+import os
+
+import paddle.inference as paddle_infer
+from paddle.fluid.core import AnalysisConfig
+from paddle.fluid.core import create_paddle_predictor
+
+from paddlenlp.transformers import ElectraTokenizer
+
+
+def parse_args():
```
Comment (ZeyuChen) [inline examples/language_model/electra/deploy/python/predict.py:13]: 为了减少args这地方占用的代码行数
建议使用#yapf: disable 和 #yapf: enable的block隔离出来
改为单行的方式来呈现

File: examples/language_model/electra/deploy/python/predict.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    #import pdb; pdb.set_trace()
+    print("inference total %s sentences done, total time : %s s" %
+          (i, time.time() - start_time))
+
+
+if __name__ == "__main__":
+    args = parse_args()
+    sentences = args.predict_sentences
+    paths = args.predict_file
+    #import pdb; pdb.set_trace()
```
Comment (ZeyuChen) [inline examples/language_model/electra/deploy/python/predict.py]: 应该删除无用的代码。
建议使用SST-2最容易理解的情感分类数据作为部署预测示例比较好

File: examples/language_model/electra/README.md (hunk 8 above)
Comment (ZeyuChen) [reply examples/language_model/electra/README.md]: 可以考虑使用GLUE/SST-2情感分类的数据集作为部署推理的示例更好。
这个是主流的NLP场景应用

File: examples/language_model/electra/README.md (hunk 4 above)
Comment (ZeyuChen) [reply examples/language_model/electra/README.md]: 建议Finetune这里直接使用简单的SST-2任务来示例，方便与后面的预测推理部署打通。

**Review by ZeyuChen — CHANGES_REQUESTED** [review]
Finetune示例建议采用SST-2，同时推理部署示例采用文本分类

File: paddlenlp/transformers/electra/tokenizer.py (hunk 5 above)
Comment (jeff41404) [reply paddlenlp/transformers/electra/tokenizer.py:124]: 已添加

File: examples/language_model/electra/README.md (hunk 4 above)
Comment (jeff41404) [reply examples/language_model/electra/README.md]: 因为预训练、Fine-tuning 和 推理部署均在不同的目录，所以在文档第3章已说明：如下所有命令均假设在 PaddleNLP/examples/language_model/electra/ 下执行，包括预训练、Fine-tuning 和 推理部署的命令。且这些相对路径均已手工验证过

File: examples/language_model/electra/README.md (hunk 6 above)
Comment (jeff41404) [reply examples/language_model/electra/README.md]: 之前是参考bert写的，现已去掉

File: examples/language_model/electra/README.md (hunk 7 above)
Comment (jeff41404) [reply examples/language_model/electra/README.md]: 已修改预测部署的input_model_dir为finetune后的模型

File: examples/language_model/electra/README.md (hunk 8 above)
Comment (jeff41404) [reply examples/language_model/electra/README.md]: 推理部署的示例已修改为SST-2

File: examples/language_model/electra/deploy/python/predict.py (hunk 10 above)
Comment (jeff41404) [reply examples/language_model/electra/deploy/python/predict.py]: 已删除

File: examples/language_model/electra/README.md (hunk 4 above)
Comment (jeff41404) [reply examples/language_model/electra/README.md]: 按照讨论，从长远考虑，已建立electra/run_glue.py，readme中相关路径已修改

File: examples/language_model/electra/deploy/python/predict.py (hunk 9 above)
Comment (jeff41404) [reply examples/language_model/electra/deploy/python/predict.py:13]: 占了较多的行是pre-commit修改，已尽量精简，目前只有9个参数

File: examples/language_model/electra/deploy/python/predict.py [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+        sen = np.array(sen).astype("int64")
+        # get input name
+        input_names = predictor.get_input_names()
+        # get input pointer and copy data
+        input_tensor = predictor.get_input_tensor(input_names[0])
+        input_tensor.reshape(sen.shape)
+        input_tensor.copy_from_cpu(sen)
+        #input_tensor.copy_from_cpu(fake_input.copy())
+
+        # run 运行predictor
```
Comment (ZeyuChen) [inline examples/language_model/electra/deploy/python/predict.py]: 不要在代码中出现中文字符。注释尽量使用全英文，否则容易在windows场景中运行python遇到中文编码问题。

File: examples/language_model/electra/export_model.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+import json
+
+import paddle
+import paddle.nn as nn
+from paddle.static import InputSpec
+
+from paddlenlp.transformers import ElectraForTotalPretraining, ElectraDiscriminator, ElectraGenerator, ElectraModel
+from paddlenlp.transformers import ElectraForSequenceClassification, ElectraTokenizer
+
+MODEL_CLASSES = {"electra": (ElectraForTotalPretraining, ElectraTokenizer), }
```
Comment (ZeyuChen) [inline examples/language_model/electra/export_model.py]: 这一变量是否已经无用了？

File: examples/language_model/electra/run_glue.py [hunk 13]
```diff
@@ -129,6 +257,12 @@ def _concat_seqs(seqs, separators, seq_mask=0, separator_mask=1):
...
 
     # Tokenize raw text
+    if len(example) == 1:
+        example = tokenizer(example[0], max_seq_len=max_seq_length)
+    else:
+        example = tokenizer(
+            example[0], text_pair=example[1], max_seq_len=max_seq_length)
+    '''
     tokens_raw = [tokenizer(l) for l in example]
     # Truncate to the truncate_length,
```
Comment (ZeyuChen) [inline examples/language_model/electra/run_glue.py]: 下面这段注释是否是有意义的？

File: examples/language_model/electra/run_pretrain.py [hunk 14]
```diff
@@ -486,49 +516,127 @@ def do_train(args):
...
+                            os.path.join(output_dir, "run_states.json"),
+                            'w') as f:
+                        json.dump(run_states, f)
                     paddle.save(model.state_dict(),
                                 os.path.join(output_dir,
                                              "model_state.pdparams"))
                     tokenizer.save_pretrained(output_dir)
                     paddle.save(optimizer.state_dict(),
                                 os.path.join(output_dir, "model_state.pdopt"))
+                    if len(log_list) > 0:
```
Comment (ZeyuChen) [inline examples/language_model/electra/run_pretrain.py:634]: 为什么要使用这种方法额外存log，而不使用户自己在外自己用重定向的方式保存log？

File: paddlenlp/transformers/electra/modeling.py [hunk 15]
```diff
@@ -697,26 +707,45 @@ def __init__(self, vocab_size, gen_weight, disc_weight):
...
 
     def forward(self, generator_prediction_scores,
                 discriminator_prediction_scores, generator_labels,
-                discriminator_labels):
+                discriminator_labels, attention_mask):
         # generator loss
         gen_loss = self.gen_loss_fct(
             paddle.reshape(generator_prediction_scores, [-1, self.vocab_size]),
             paddle.reshape(generator_labels, [-1]))
         # todo: we can remove 4 lines after when CrossEntropyLoss(reduction='mean') improved
```
Comment (ZeyuChen) [inline paddlenlp/transformers/electra/modeling.py:719]: 这一个TODO最新的Paddle版本仍然是这样吗？

File: examples/language_model/electra/run_glue.py [hunk 16]
```diff
@@ -30,11 +29,16 @@
...
 from paddlenlp.data import Stack, Tuple, Pad
 from paddlenlp.data.sampler import SamplerHelper
+from paddlenlp.transformers import BertForSequenceClassification, BertTokenizer
 from paddlenlp.transformers import ElectraForSequenceClassification, ElectraTokenizer
+from paddlenlp.transformers import ErnieForSequenceClassification, ErnieTokenizer
 from paddlenlp.transformers import LinearDecayWithWarmup
-from paddlenlp.utils.log import logger
 from paddlenlp.metrics import AccuracyAndF1, Mcc, PearsonAndSpearman
 
+FORMAT = '%(asctime)s-%(levelname)s: %(message)s'
```
Comment (ZeyuChen) [inline examples/language_model/electra/run_glue.py:38]: 这里的logger单独拉3行代码出来的目的是？

File: examples/language_model/electra/run_glue.py [hunk 17]
```diff
@@ -47,10 +51,117 @@
 }
 
 MODEL_CLASSES = {
```
Comment (ZeyuChen) [inline examples/language_model/electra/run_glue.py:53]: 如果这个目录专门针对electra的话，我们是否还需要用MODEL_CLASSES的方式来管理？

File: examples/language_model/electra/run_glue.py [hunk 18]
```diff
@@ -47,10 +51,117 @@
...
+
+    # Required parameters
+    parser.add_argument(
+        "--task_name",
+        default=None,
+        type=str,
+        required=True,
+        help="The name of the task to train selected in the list: " +
+        ", ".join(TASK_CLASSES.keys()), )
+    parser.add_argument(
```
Comment (ZeyuChen) [inline examples/language_model/electra/run_glue.py:71]: 这一args是否可以删除？

**Review by ZeyuChen — COMMENTED** [review]
留了一些意见

File: examples/language_model/electra/deploy/python/predict.py (hunk 11 above)
Comment (jeff41404) [reply examples/language_model/electra/deploy/python/predict.py]: 已去除中文

File: examples/language_model/electra/export_model.py (hunk 12 above)
Comment (jeff41404) [reply examples/language_model/electra/export_model.py]: 原来考虑用于export pretrain model，目前主要是export fine-tune model，已去除

File: examples/language_model/electra/run_glue.py (hunk 13 above)
Comment (jeff41404) [reply examples/language_model/electra/run_glue.py]: 这里应该是之前其它同学关于tokenizer升级的改动，已删除

File: examples/language_model/electra/run_pretrain.py (hunk 14 above)
Comment (jeff41404) [reply examples/language_model/electra/run_pretrain.py:634]: 这样log能够同时显示在终端和文件中，huggingface也是类似实现

File: paddlenlp/transformers/electra/modeling.py (hunk 15 above)
Comment (jeff41404) [reply paddlenlp/transformers/electra/modeling.py:719]: 是的，还没有完全修复

File: examples/language_model/electra/run_glue.py (hunk 16 above)
Comment (jeff41404) [reply examples/language_model/electra/run_glue.py:38]: 如沟通，这些diff是先删除run_glue再从glue/run_glue导致的，其实本次重点是预训练和fine-tuning部分，没有修改run_glue

File: examples/language_model/electra/run_glue.py (hunk 17 above)
Comment (jeff41404) [reply examples/language_model/electra/run_glue.py:53]: 如沟通，这些diff是先删除run_glue再从glue/run_glue导致的，其实本次重点是预训练和fine-tuning部分，没有修改run_glue，之后可以再考虑精简

File: examples/language_model/electra/run_glue.py (hunk 18 above)
Comment (jeff41404) [reply examples/language_model/electra/run_glue.py:71]: 如沟通，这些diff是先删除run_glue再从glue/run_glue导致的，其实本次重点是预训练和fine-tuning部分，没有修改run_glue

**Review by ZeyuChen — APPROVED** [review]
LGTM
